scheds/rust/scx_layered/src/bpf/cost.bpf.c

```diff
@@ -102,39 +120,71 @@ static struct cost *initialize_cost(u32 cost_idx, u32 parent_idx,
 }
 
 /*
- * Initializes the cost of a layer.
+ * Initializes a budget.
  */
-static void initialize_cost_layer(struct cost *costc, u32 layer_id, s64 capacity)
+static void initialize_budget(struct cost *costc, u32 budget_id, s64 capacity)
 {
-	costc->capacity[layer_id] = capacity;
-	costc->budget[layer_id] = capacity;
+	if (budget_id >= MAX_GLOBAL_BUDGETS) {
+		scx_bpf_error("invalid budget id %d", budget_id);
+		return;
+	}
+	costc->capacity[budget_id] = capacity;
+	costc->budget[budget_id] = capacity;
 }
 
 /*
- * Returns the preferred layer based on the layer with the maximum budget.
+ * Calculates the preferred budget and layer based based on maximum budget.
  */
-static u32 preferred_cost(struct cost *costc)
+static void calc_preferred_cost(struct cost *costc)
 {
-	u32 layer_id, id, max_layer = 0;
+	u32 layer_id, id, budget_id, pref_budget = 0, max_layer = 0;
 	s64 max_budget = 0;
+	u64 dsq_id;
 	u32 rotation = bpf_get_smp_processor_id() % nr_layers;
 
 	bpf_for(id, 0, nr_layers) {
-		// If there is two equally weighted layers that have the same
-		// budget we rely on rotating the layers based on the cpu. This
-		// may not work well on low core machines.
+		/* 
+		 * If there is two equally weighted layers that have the same
+		 * budget we rely on rotating the layers based on the cpu. This
+		 * may not work well on low core machines.
+		 */
 		layer_id = rotate_layer_id(id, rotation);
 		if (layer_id > nr_layers) {
 			scx_bpf_error("invalid layer");
-			return 0;
+			return;
 		}
 		if (costc->budget[layer_id] > max_budget) {
 			max_budget = costc->budget[layer_id];
 			max_layer = layer_id;
+			pref_budget = max_layer;
 		}
 	}
+	// Hi fallback DSQs
+	bpf_for(id, 0, nr_llcs) {
+		dsq_id = llc_hi_fallback_dsq_id(id);
+		budget_id = fallback_dsq_cost_id(dsq_id);
```

Comment:
Don't these cancel each other out in this scenario? The first one always adds ```HI_FALLBACK_DSQ_BASE```, and the other always subtracts it. Can we just use ```id``` here instead of ```budget_id``` in the loop then? This also allows us to remove the check in 166-168.

Reply:
I suppose so, I just worry if that ever changes it might break but good catch!